src/mlir/generator.cc

```diff
@@ -263,11 +264,12 @@ namespace mlir::verona
       return builder.getNoneType();
 
     // If type is in the alias table, get it
-    if (typeTable.inScope(desc))
-      return typeTable.lookup(desc);
+    auto type = typeTable.lookup(desc);
```

Comment:
Can this use operator[] and assign to the result on line 273, rather than doing the search twice?

Reply:
I'm trying to avoid the call to `getOpaqueType()`. Initially, types will search twice, but after the types are all declared, they'll only be searched once (to retrieve from the cache).

If I switch the order, I need to generate the opaque type, and then every call, including when it's in the cache, will create a new type and then realise it's already there, and dispose of the temp.

Reply:
I meant something along the lines of:
```c++
auto &type = typeTable[desc];
if (!type)
  type = getOpaqueType(desc, context)
return type;
```
Would this not work?

Reply:
First, this is a cache pattern. I need to insert it into the table before returning. I can't return straight from `getOpaqueType`.

Second, the table isn't a simple map, it's a stack of maps. There are two searches: `inScope` only looks at the current lexical frame while `lookup` looks at the whole stack. Overloading `[]` could be done to either of them and it's not clear to me which would make more sense.

Finally, before, I was doing a local search ('inScope') and then returning it only if it was in the current scope (not the lexical parents), which is wrong. I need to traverse all contexts. This is an expensive search for new types, but so is creating the type (it will get more complex in the future). The idea is that we'll use a lot more types than we'll create new ones.

Reply:
I have just pushed a new commit that considerably simplifies the symbol table. It was a hybrid of a few past attempts and extremely redundant. I think this fix and the new implementation will considerably reduce the complexity.

Reply:
I see.  If you miss in the first scope, you need to search the others anyway so that you can insert.  That can lead to quite expensive lookups for nested scopes, but it's probably fine.  In the past, I've implemented this with a map that does CoW for the buckets, so creating a new scope is a cheap refcounted 'copy', inserting new symbols may be relatively expensive (actually copies part of the map) and lookup cost is independent of nesting complexity and the lookup-and-insert-if-not-present operation is cheaper than lookup then insert.  

I don't think that's necessary here.